// @mui/joy/node/Switch/Switch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _clsx = _interopRequireDefault(require("clsx"));

var _utils = require("@mui/utils");

var _base = require("@mui/base");

var _SwitchUnstyled = require("@mui/base/SwitchUnstyled");

var _styles = require("../styles");

var _switchClasses = _interopRequireWildcard(require("./switchClasses"));

var _jsxRuntime = require("react/jsx-runtime");

const _excluded = ["checked", "className", "component", "componentsProps", "defaultChecked", "disabled", "onBlur", "onChange", "onFocus", "onFocusVisible", "readOnly", "required", "id", "color", "variant", "size", "startDecorator", "endDecorator"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const useUtilityClasses = ownerState => {
  const {
    checked,
    disabled,
    focusVisible,
    readOnly,
    color,
    variant
  } = ownerState;
  const slots = {
    root: ['root', checked && 'checked', disabled && 'disabled', focusVisible && 'focusVisible', readOnly && 'readOnly', variant && `variant${(0, _utils.unstable_capitalize)(variant)}`, color && `color${(0, _utils.unstable_capitalize)(color)}`],
    thumb: ['thumb', checked && 'checked'],
    track: ['track', checked && 'checked'],
    action: ['action', focusVisible && 'focusVisible'],
    input: ['input']
  };
  return (0, _base.unstable_composeClasses)(slots, _switchClasses.getSwitchUtilityClass, {});
};

const switchColorVariables = ({
  theme,
  ownerState
}) => (data = {}) => {
  var _theme$vars$palette, _theme$vars$palette2, _theme$vars$palette3, _theme$vars$palette4, _theme$vars$palette5;

  const variant = ownerState.variant;
  const color = ownerState.color;
  return {
    '--Switch-track-background': (_theme$vars$palette = theme.vars.palette[color]) == null ? void 0 : _theme$vars$palette[`${variant}${data.state || ''}Bg`],
    '--Switch-track-color': (_theme$vars$palette2 = theme.vars.palette[color]) == null ? void 0 : _theme$vars$palette2[`${variant}Color`],
    '--Switch-track-borderColor': variant === 'outlined' ? (_theme$vars$palette3 = theme.vars.palette[color]) == null ? void 0 : _theme$vars$palette3[`${variant}${data.state || ''}Border`] : 'currentColor',
    '--Switch-thumb-background': (_theme$vars$palette4 = theme.vars.palette[color]) == null ? void 0 : _theme$vars$palette4[`${variant}${data.state || ''}Color`],
    '--Switch-thumb-color': (_theme$vars$palette5 = theme.vars.palette[color]) == null ? void 0 : _theme$vars$palette5[`${variant}Bg`]
  };
};

const SwitchRoot = (0, _styles.styled)('span', {
  name: 'JoySwitch',
  slot: 'Root',
  overridesResolver: (props, styles) => styles.root
})(({
  theme,
  ownerState
}) => {
  var _theme$variants, _theme$variants2;

  const getColorVariables = switchColorVariables({
    theme,
    ownerState
  });
  return (0, _extends2.default)({
    '--variant-borderWidth': (_theme$variants = theme.variants[ownerState.variant]) == null ? void 0 : (_theme$variants2 = _theme$variants[ownerState.color]) == null ? void 0 : _theme$variants2['--variant-borderWidth'],
    '--Switch-track-radius': theme.vars.radius.lg,
    '--Switch-thumb-shadow': ownerState.variant === 'soft' ? 'none' : '0 0 0 1px var(--Switch-track-background)'
  }, ownerState.size === 'sm' && {
    '--Switch-track-width': '40px',
    '--Switch-track-height': '20px',
    '--Switch-thumb-size': '12px',
    '--Switch-gap': '6px',
    fontSize: theme.vars.fontSize.sm
  }, ownerState.size === 'md' && {
    '--Switch-track-width': '48px',
    '--Switch-track-height': '24px',
    '--Switch-thumb-size': '16px',
    '--Switch-gap': '8px',
    fontSize: theme.vars.fontSize.md
  }, ownerState.size === 'lg' && {
    '--Switch-track-width': '64px',
    '--Switch-track-height': '32px',
    '--Switch-thumb-size': '24px',
    '--Switch-gap': '12px'
  }, {
    '--internal-paddingBlock': `max((var(--Switch-track-height) - 2 * var(--variant-borderWidth) - var(--Switch-thumb-size)) / 2, 0px)`,
    '--Switch-thumb-radius': `max((var(--Switch-track-radius) - var(--variant-borderWidth)) - var(--internal-paddingBlock), min(var(--internal-paddingBlock) / 2, (var(--Switch-track-radius) - var(--variant-borderWidth)) / 2))`,
    '--Switch-thumb-width': 'var(--Switch-thumb-size)',
    '--Switch-thumb-offset': `max((var(--Switch-track-height) - var(--Switch-thumb-size)) / 2, 0px)`
  }, getColorVariables(), {
    '&:hover': (0, _extends2.default)({}, getColorVariables({
      state: 'Hover'
    })),
    [`&.${_switchClasses.default.checked}`]: (0, _extends2.default)({}, getColorVariables(), {
      '&:hover': (0, _extends2.default)({}, getColorVariables({
        state: 'Hover'
      }))
    }),
    [`&.${_switchClasses.default.disabled}`]: (0, _extends2.default)({
      pointerEvents: 'none',
      color: theme.vars.palette.text.tertiary
    }, getColorVariables({
      state: 'Disabled'
    })),
    display: 'inline-flex',
    alignItems: 'center',
    alignSelf: 'center',
    fontFamily: theme.vars.fontFamily.body,
    position: 'relative',
    padding: 'calc((var(--Switch-thumb-size) / 2) - (var(--Switch-track-height) / 2)) calc(-1 * var(--Switch-thumb-offset))',
    backgroundColor: 'initial',
    // clear background in case `outlined` variant contain background.
    border: 'none'
  });
});
const SwitchAction = (0, _styles.styled)('div', {
  name: 'JoySwitch',
  slot: 'Action',
  overridesResolver: (props, styles) => styles.action
})(({
  theme
}) => ({
  borderRadius: 'var(--Switch-track-radius)',
  position: 'absolute',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
  [theme.focus.selector]: theme.focus.default
}));
const SwitchInput = (0, _styles.styled)('input', {
  name: 'JoySwitch',
  slot: 'Input',
  overridesResolver: (props, styles) => styles.input
})({
  margin: 0,
  height: '100%',
  width: '100%',
  opacity: 0,
  position: 'absolute',
  cursor: 'pointer'
});
const SwitchTrack = (0, _styles.styled)('span', {
  name: 'JoySwitch',
  slot: 'Track',
  overridesResolver: (props, styles) => styles.track
})(({
  theme,
  ownerState
}) => (0, _extends2.default)({
  position: 'relative',
  color: 'var(--Switch-track-color)',
  height: 'var(--Switch-track-height)',
  width: 'var(--Switch-track-width)',
  display: 'flex',
  flexShrink: 0,
  justifyContent: 'space-between',
  alignItems: 'center',
  boxSizing: 'border-box',
  border: 'var(--variant-borderWidth) solid',
  borderColor: 'var(--Switch-track-borderColor)',
  backgroundColor: 'var(--Switch-track-background)',
  borderRadius: 'var(--Switch-track-radius)',
  fontFamily: theme.vars.fontFamily.body
}, ownerState.size === 'sm' && {
  fontSize: theme.vars.fontSize.xs
}, ownerState.size === 'md' && {
  fontSize: theme.vars.fontSize.sm
}, ownerState.size === 'lg' && {
  fontSize: theme.vars.fontSize.md
}));
const SwitchThumb = (0, _styles.styled)('span', {
  name: 'JoySwitch',
  slot: 'Thumb',
  overridesResolver: (props, styles) => styles.thumb
})({
  '--Icon-fontSize': 'calc(var(--Switch-thumb-size) * 0.75)',
  transition: 'left 0.2s',
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  position: 'absolute',
  top: '50%',
  left: 'calc(50% - var(--Switch-track-width) / 2 + var(--Switch-thumb-width) / 2 + var(--Switch-thumb-offset))',
  transform: 'translate(-50%, -50%)',
  width: 'var(--Switch-thumb-width)',
  height: 'var(--Switch-thumb-size)',
  borderRadius: 'var(--Switch-thumb-radius)',
  boxShadow: 'var(--Switch-thumb-shadow)',
  color: 'var(--Switch-thumb-color)',
  backgroundColor: 'var(--Switch-thumb-background)',
  [`&.${_switchClasses.default.checked}`]: {
    left: 'calc(50% + var(--Switch-track-width) / 2 - var(--Switch-thumb-width) / 2 - var(--Switch-thumb-offset))'
  }
});
const SwitchStartDecorator = (0, _styles.styled)('span', {
  name: 'JoySwitch',
  slot: 'StartDecorator',
  overridesResolver: (props, styles) => styles.startDecorator
})({
  display: 'inline-flex',
  marginInlineEnd: 'var(--Switch-gap)'
});
const SwitchEndDecorator = (0, _styles.styled)('span', {
  name: 'JoySwitch',
  slot: 'EndDecorator',
  overridesResolver: (props, styles) => styles.endDecorator
})({
  display: 'inline-flex',
  marginInlineStart: 'var(--Switch-gap)'
});
const Switch = /*#__PURE__*/React.forwardRef(function Switch(inProps, ref) {
  var _componentsProps$star, _componentsProps$trac, _componentsProps$trac2, _componentsProps$thum, _componentsProps$acti, _componentsProps$inpu, _componentsProps$endD;

  const props = inProps;
  const {
    checked: checkedProp,
    className,
    component,
    componentsProps = {},
    defaultChecked,
    disabled: disabledProp,
    onBlur,
    onChange,
    onFocus,
    onFocusVisible,
    readOnly: readOnlyProp,
    id,
    color,
    variant = 'solid',
    size = 'md',
    startDecorator,
    endDecorator
  } = props,
        otherProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const useSwitchProps = {
    checked: checkedProp,
    defaultChecked,
    disabled: disabledProp,
    onBlur,
    onChange,
    onFocus,
    onFocusVisible,
    readOnly: readOnlyProp
  };
  const {
    getInputProps,
    checked,
    disabled,
    focusVisible,
    readOnly
  } = (0, _SwitchUnstyled.useSwitch)(useSwitchProps);
  const ownerState = (0, _extends2.default)({}, props, {
    id,
    checked,
    disabled,
    focusVisible,
    readOnly,
    color: checked ? color || 'primary' : color || 'neutral',
    variant,
    size
  });
  const classes = useUtilityClasses(ownerState);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(SwitchRoot, (0, _extends2.default)({
    ref: ref
  }, otherProps, {
    as: component,
    ownerState: ownerState,
    className: (0, _clsx.default)(classes.root, className),
    children: [startDecorator && /*#__PURE__*/(0, _jsxRuntime.jsx)(SwitchStartDecorator, (0, _extends2.default)({
      "aria-hidden": "true"
    }, componentsProps.startDecorator, {
      ownerState: ownerState,
      className: (0, _clsx.default)(classes.input, (_componentsProps$star = componentsProps.startDecorator) == null ? void 0 : _componentsProps$star.className),
      children: typeof startDecorator === 'function' ? startDecorator(ownerState) : startDecorator
    })), /*#__PURE__*/(0, _jsxRuntime.jsxs)(SwitchTrack, (0, _extends2.default)({}, componentsProps.track, {
      ownerState: ownerState,
      className: (0, _clsx.default)(classes.track, (_componentsProps$trac = componentsProps.track) == null ? void 0 : _componentsProps$trac.className),
      children: [(_componentsProps$trac2 = componentsProps.track) == null ? void 0 : _componentsProps$trac2.children, /*#__PURE__*/(0, _jsxRuntime.jsx)(SwitchThumb, (0, _extends2.default)({}, componentsProps.thumb, {
        ownerState: ownerState,
        className: (0, _clsx.default)(classes.thumb, (_componentsProps$thum = componentsProps.thumb) == null ? void 0 : _componentsProps$thum.className)
      }))]
    })), /*#__PURE__*/(0, _jsxRuntime.jsx)(SwitchAction, (0, _extends2.default)({}, componentsProps.action, {
      ownerState: ownerState,
      className: (0, _clsx.default)(classes.action, (_componentsProps$acti = componentsProps.action) == null ? void 0 : _componentsProps$acti.className),
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(SwitchInput, (0, _extends2.default)({
        id: id
      }, componentsProps.input, {
        ownerState: ownerState
      }, getInputProps(), {
        className: (0, _clsx.default)(classes.input, (_componentsProps$inpu = componentsProps.input) == null ? void 0 : _componentsProps$inpu.className)
      }))
    })), endDecorator && /*#__PURE__*/(0, _jsxRuntime.jsx)(SwitchEndDecorator, (0, _extends2.default)({
      "aria-hidden": "false"
    }, componentsProps.endDecorator, {
      ownerState: ownerState,
      className: (0, _clsx.default)(classes.input, (_componentsProps$endD = componentsProps.endDecorator) == null ? void 0 : _componentsProps$endD.className),
      children: typeof endDecorator === 'function' ? endDecorator(ownerState) : endDecorator
    }))]
  }));
});
process.env.NODE_ENV !== "production" ? Switch.propTypes
/* remove-proptypes */
= {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // |     To update them edit TypeScript types and run "yarn proptypes"  |
  // ----------------------------------------------------------------------

  /**
   * If `true`, the component is checked.
   */
  checked: _propTypes.default.bool,

  /**
   * @ignore
   */
  children: _propTypes.default.node,

  /**
   * Class name applied to the root element.
   */
  className: _propTypes.default.string,

  /**
   * The color of the component. It supports those theme colors that make sense for this component.
   * @default 'neutral'
   */
  color: _propTypes.default
  /* @typescript-to-proptypes-ignore */
  .oneOfType([_propTypes.default.oneOf(['danger', 'info', 'primary', 'success', 'warning']), _propTypes.default.string]),

  /**
   * The component used for the Root slot.
   * Either a string to use a HTML element or a component.
   */
  component: _propTypes.default.elementType,

  /**
   * The props used for each slot inside the Switch.
   * @default {}
   */
  componentsProps: _propTypes.default.shape({
    action: _propTypes.default.object,
    endDecorator: _propTypes.default.object,
    input: _propTypes.default.object,
    startDecorator: _propTypes.default.object,
    thumb: _propTypes.default.object,
    track: _propTypes.default.object
  }),

  /**
   * The default checked state. Use when the component is not controlled.
   */
  defaultChecked: _propTypes.default.bool,

  /**
   * If `true`, the component is disabled.
   */
  disabled: _propTypes.default.bool,

  /**
   * The element that appears at the end of the switch.
   */
  endDecorator: _propTypes.default
  /* @typescript-to-proptypes-ignore */
  .oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * @ignore
   */
  id: _propTypes.default.string,

  /**
   * @ignore
   */
  onBlur: _propTypes.default.func,

  /**
   * Callback fired when the state is changed.
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event The event source of the callback.
   * You can pull out the new value by accessing `event.target.value` (string).
   * You can pull out the new checked state by accessing `event.target.checked` (boolean).
   */
  onChange: _propTypes.default.func,

  /**
   * @ignore
   */
  onFocus: _propTypes.default.func,

  /**
   * @ignore
   */
  onFocusVisible: _propTypes.default.func,

  /**
   * If `true`, the component is read only.
   */
  readOnly: _propTypes.default.bool,

  /**
   * If `true`, the `input` element is required.
   */
  required: _propTypes.default.bool,

  /**
   * The size of the component.
   * @default 'md'
   */
  size: _propTypes.default
  /* @typescript-to-proptypes-ignore */
  .oneOfType([_propTypes.default.oneOf(['sm', 'md', 'lg']), _propTypes.default.string]),

  /**
   * The element that appears at the end of the switch.
   */
  startDecorator: _propTypes.default
  /* @typescript-to-proptypes-ignore */
  .oneOfType([_propTypes.default.node, _propTypes.default.func]),

  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object]),

  /**
   * The variant to use.
   * @default 'solid'
   */
  variant: _propTypes.default
  /* @typescript-to-proptypes-ignore */
  .oneOfType([_propTypes.default.oneOf(['outlined', 'plain', 'soft', 'solid']), _propTypes.default.string])
} : void 0;
var _default = Switch;
exports.default = _default;
